blogposts/2021/how-we-built-batch-changes.md

```diff
@@ -0,0 +1,202 @@
+container we could give them a place to plug it in — making it large-scale.
+
+This radical change to how the product worked, was, in my opinion, maybe the
+most important one made in the life of Batch Changes. The mental model and the
+abstraction layers it introduced turned out to be incredibly well-fitting. They
+allowed us to go faster and get a _valuable_ tool into the hands of our
+customers even earlier.
+
+So far, so good, right? Yes, except, you know, [naming is
+hard](https://martinfowler.com/bliki/TwoHardThings.html) and while we could've
+gone with the old name for the feature — _Automation_ — we decided to rename it.
+Take a look at [PR #3][p3] for the first of many in February 2020 that changed
+"[Aa]utomation" to "[Cc]ampaigns" in our codebase.
+
+What followed were months of fine-tuning, user testing, shipping, tweaking -
+_iterating_. We added more features, fixed a lot of bugs, made things go faster.
+
+But — and by now you should know where this is going — something felt _off_.
+While Campaigns worked and we had customers using it and saying they loved it,
+we also noticed that our colleagues had problems using it, often hesitating to
+admit that they were confused with the workflow. Some of the questions we got:
+why JSON (yes, we made users write JSON by hand)? Why can't I put everything in
+that JSON file, why do I have to switch to the web UI? And what if I want to
+update the changes?
+
+I wouldn't say we were stuck again - we were constantly shipping new features
+and improvements. But it also started dawning on us that maybe we've exhausted
+one idea. Tweaking wouldn't get us out of this, nor would a sledgehammer. But maybe if
+we look at it from a different angle? What if, instead of requiring users to say
+what should happen to produce a campaign, we switch to a _declarative_ model and
+let them describe what the campaign should like in a YAML file?
+
+[This pull request](https://github.com/sourcegraph/sourcegraph/pull/10921) shows
+how we asked "what if?": we wrote the documentation first (yes, README-driven
+development) and showed it to teammates and colleagues. The time it took a
+colleague to go from "ok, tell me what campaigns are" to "ohhh, I get it! nice"
+was reduced tenfold. Very much to our relief, since the changes necessary to
+make the documentation a reality were huge. We had to build a distributed,
+_declarative_ system that manages hundreds or thousands of pull and merge
+requests across different code hosts.
+
+Once the idea was validated by a lot of "this totally makes sense" from potential
+users, we got to work. Look at [this PR][pr4b], number 4b in this series if you
+will, to see an example of what "declarative" means and how we reconcile the
+current and the desired state (as described by a user) of a changeset. (Sidenote
+for the curious: the code today is [even cooler][evencoolercode], take a look).
+
+That brings us to the last of the five pull requests.
+
```

Comment:
```suggestion
## Pull request 5: Main point of this PR reference
```

Reply:
I have to admit that it took me really long to understand that "Main point of this PR reference" was directed at me 😄